### PR TITLE
Fix issues with serializing BulkByScrollResponse

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently there are two issues with serializing `BulkByScrollResponse`.
First, when deserializing from XContent, indexing exceptions and search
exceptions are switched. Additionally, search exceptions do no retain
the appropriate RestStatus code, so you must evaluate the status code
from the exception. However, the exception class is not always correctly
retained when serialized.

This commit adds tests in the failure case. Additionally, fixes the
swapping of failure types and adds the rest status code to the search
failure.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/reindex/BulkByScrollResponseTests.java [hunk 1]
```diff
@@ -137,19 +143,26 @@ protected void assertEqualInstances(BulkByScrollResponse expected, BulkByScrollR
...
     }
 
     @Override
     protected BulkByScrollResponse doParseInstance(XContentParser parser) throws IOException {
         return BulkByScrollResponse.fromXContent(parser);
     }
 
+    @Override
+    protected boolean assertToXContentEquivalence() {
+        // XContentEquivalence fails in the exception case, due to how exceptions are serialized.
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/reindex/BulkByScrollResponseTests.java]: I think this means that
toXContext(x) != toXContent(fromXContent(toXContent(x)))
such that we loose more info on the second toXContent? I would think that we would like to keep this check, is there a good reason why we cannot do this?

**Review by henningandersen — COMMENTED** [review]
I left a few smaller comments to address, looking good otherwise.

File: server/src/test/java/org/elasticsearch/index/reindex/BulkByScrollResponseTests.java [hunk 2]
```diff
@@ -126,7 +132,7 @@ public static void assertEqualBulkResponse(BulkByScrollResponse expected, BulkBy
             assertEquals(expectedFailure.getIndex(), actualFailure.getIndex());
             assertEquals(expectedFailure.getShardId(), actualFailure.getShardId());
             assertEquals(expectedFailure.getNodeId(), actualFailure.getNodeId());
-            assertThat(expectedFailure.getReason().getMessage(), containsString(actualFailure.getReason().getMessage()));
```
[the remark this answers is not in the document]
Comment (tbrooks8) [reply server/src/test/java/org/elasticsearch/index/reindex/BulkByScrollResponseTests.java]: We can kind of keep this check, but I had to switch it around. The serialization adds the exception type to the message.

File: server/src/test/java/org/elasticsearch/index/reindex/BulkByScrollResponseTests.java (hunk 1 above)
Comment (tbrooks8) [reply server/src/test/java/org/elasticsearch/index/reindex/BulkByScrollResponseTests.java]: Yes serializing exceptions multiple times is not equivalent. You serialize resource not found exception, and you lose the resource not found on the deserialization. So then the next time you serialize it, it is represented differently. 

I should point out I have maintained this check in the success case (which is the only case tested prior to this PR).

**Comment by tbrooks8** [comment]
Here is how a `NoNodeAvailableException` changes:

First serialization:
{reason=baz, type=no_node_available_exception}

Second serialization:
{reason=Elasticsearch exception [type=no_node_available_exception, reason=baz], type=exception}

Here is how a `ElasticsearchException` changes:

First serialization:
{reason=baz, type=exception}

Second serialization:
{reason=Elasticsearch exception [type=exception, reason=baz], type=exception}

**Comment by tbrooks8** [comment]
It might be possible to parse the `type` in a manner that is BWC. It's something that we can look into as a more general follow-up.

**Review by henningandersen — APPROVED** [review]
LGTM.

Thanks for the extra info on the exception serialization to xcontent. I think we should be able to deserialize into a special exception type that holds the reason and type fields and is serialized to the held reason and type. But let us leave that exercise for a follow-up.

**Pull request merged**